src/Features/Core/Portable/ConvertNumericLiteral/AbstractConvertNumericLiteralCodeRefactoringProvider.cs

```diff
@@ -4,169 +4,134 @@
-            }
-            else
-            {
-                switch (kind)
-                {
-                    case NumericKind.Decimal when number.Length > 3:
-                        RegisterRefactoringWithResult(AddSeparators(number, interval: 3), FeaturesResources.Separate_thousands);
-                        break;
-
-                    case NumericKind.Hexadecimal when number.Length > 4:
-                        RegisterRefactoringWithResult(hexPrefix + AddSeparators(number, interval: 4), FeaturesResources.Separate_words);
-                        break;
-
-                    case NumericKind.Binary when number.Length > 4:
-                        RegisterRefactoringWithResult(binaryPrefix + AddSeparators(number, interval: 4), FeaturesResources.Separate_nibbles);
-                        break;
-                }
-            }
-
-            void RegisterRefactoringWithResult(string text, string title)
-            {
-                context.RegisterRefactoring(
-                    CodeAction.Create(title, c => ReplaceTokenAsync(document, root, numericToken, value, text, suffix), title),
-                    numericToken.Span);
-            }
-        }
+        if (kind != NumericKind.Hexadecimal)
+            result.Add(CreateCodeAction(hexPrefix + value.ToString("X"), FeaturesResources.Convert_to_hex));
 
-        private static Task<Document> ReplaceTokenAsync(Document document, SyntaxNode root, SyntaxToken numericToken, long value, string text, string suffix)
+        const string DigitSeparator = "_";
+        if (numericText.Contains(DigitSeparator))
         {
-            var generator = SyntaxGenerator.GetGenerator(document);
-            var updatedToken = generator.NumericLiteralToken(text + suffix, (ulong)value)
-                .WithTriviaFrom(numericToken);
-            var updatedRoot = root.ReplaceToken(numericToken, updatedToken);
-            return Task.FromResult(document.WithSyntaxRoot(updatedRoot));
+            result.Add(CreateCodeAction(prefix + number.Replace(DigitSeparator, string.Empty), FeaturesResources.Remove_separators));
         }
-
-        internal virtual async Task<SyntaxToken> GetNumericTokenAsync(CodeRefactoringContext context)
+        else
         {
-            var syntaxFacts = context.Document.GetRequiredLanguageService<ISyntaxFactsService>();
-
-            var literalNode = await context.TryGetRelevantNodeAsync<TNumericLiteralExpression>().ConfigureAwait(false);
-            var numericLiteralExpressionNode = syntaxFacts.IsNumericLiteralExpression(literalNode)
-                ? literalNode
-                : null;
-
-            return numericLiteralExpressionNode != null
-                ? numericLiteralExpressionNode.GetFirstToken()    // We know that TNumericLiteralExpression has always only one token: NumericLiteralToken
-                : default;
+            result.AsRef().AddIfNotNull(kind switch
```

Comment:
curious, why is AddIfNotNull not an instance methdo?

Reply:
because it needs to be specalized for when T is a struct vs when T is a reference type.  As an instance method, that's not possible.